lib/_http_outgoing.js

```diff
@@ -109,81 +111,187 @@ OutgoingMessage.prototype.setTimeout = function(msecs, callback) {
 // any messages, before ever calling this.  In that case, just skip
 // it, since something else is destroying this connection anyway.
 OutgoingMessage.prototype.destroy = function(error) {
-  if (this.socket)
-    this.socket.destroy(error);
-  else
-    this.once('socket', function(socket) {
+  const socket = this.socket;
+
+  if (socket) {
+    // NOTE: we may have some nextTick-corked data pending to write. There is
+    // no point in nextTick optimization here anymore, so just write everything
+    // to the socket
+    socket.uncork();
+    socket.destroy(error);
+  } else {
+    this.once('socket', (socket) => {
+      socket.uncork();
       socket.destroy(error);
     });
+  }
+};
+
+
+OutgoingMessage.prototype._sendHeaders = function(data) {
+  if (this._headerSent)
+    return;
+
+  if (!this.socket) {
+    this.once('socket', () => {
+      this._sendHeaders();
+    });
+    return;
+  }
+
+  if (!this._header)
+    this._implicitHeader();
+
+  this._headerSent = true;
+  if (this.socket.writable && !this.socket.destroyed)
+    this.socket.write(this._header);
 };
 
 
-// This abstract either writing directly to the socket or buffering it.
-OutgoingMessage.prototype._send = function(data, encoding, callback) {
-  // This is a shameful hack to get the headers and first body chunk onto
-  // the same packet. Future versions of Node are going to take care of
-  // this at a lower level and in a more general way.
+OutgoingMessage.prototype._countPendingData = function _countPendingData() {
+  if (!this._onPendingData)
+    return;
+
+  const before = this._onPendingLast;
+  const now = this._writableState.length;
+  this._onPendingLast = now;
+
+  this._onPendingData(now - before);
+};
+
+
+OutgoingMessage.prototype.write = function write(data, encoding, callback) {
+  const res = Writable.prototype.write.call(this, data, encoding, callback);
```

Comment:
The following may seem like a bit of a nasty hack, but I'll put it out here for what it's worth, in case you're worried about performance in `write()` and `end()`.

```js
OutgoingMessage.prototype._writableWrite = Writable.prototype.write;

OutgoingMessage.prototype.write = function write(data, encoding, callback) {
  this._writableWrite(data, encoding, callback);
  this._countPendingData();
  debug('write ret = %j', res);
  return res;
};
```

This avoids the need for `call()` as `this` will remain what it needs to be.